### ISSUE TITLE
UTS Tugas 1 Machine Learning Nessa 41155050210057

### ISSUE THREAD

**Issue by nessa0712** [nessa0712/UTS-Machine-Learning-Nessa-InformatikaB#1]
### TUGAS 1 
**Pendahuluan Machine Learning**

1. Instalasi Jupyter Noterbook dan Library python seperti NumPy, SciPy, Pandas, Matplotlib, Seaborn, Scikit-learn.
![Instalasi Jupyter Noterbook dan Library python seperti NumPy, SciPy, Pandas, Matplotlib, Seaborn, Scikit-learn](https://github.com/user-attachments/assets/1f509571-ac22-4cc2-89e6-79063be34deb)
_Install Jupyter Notebook_
![Install Numpy](https://github.com/user-attachments/assets/56eb2e42-b16e-42aa-b1a6-ee3904076cda)
 _Install Library Numpy_
![Install Scipy](https://github.com/user-attachments/assets/4abad4c7-d6cf-416f-a066-86eb26f28478)
_Install Library Scipy_
![Install Pandan](https://github.com/user-attachments/assets/aad02a67-1156-46d0-aa24-3c6ca1267d75)
_Install Library Pandas_
![Matplotlib](https://github.com/user-attachments/assets/f65334bb-9d9b-4f01-af57-e86ff002ff4c)
_Install Library Matplotlib_
![Seaborn](https://github.com/user-attachments/assets/f078f097-6ad3-4657-b508-ed46dde19883)
_Install Library Seaborn_
![Scikit-learn](https://github.com/user-attachments/assets/c425e3dc-1f6d-4a76-bf79-182a0c9596ac)
_Install Library Scikit-learn_
![New Worksheet](https://github.com/user-attachments/assets/af8340de-d2b6-4b10-987c-30c168968891)
_Membuat worksheet baru dengan python 3_
![Input NAma dan NPM](https://github.com/user-attachments/assets/9abcc2bc-da6b-4a2f-a8fa-b1c13c15646c)
_Menginputkan Nama dan NPM kemudian ditampilkan_

2.  Menggunakan Google Collab dan mencoba untuk menginputkan nama dan NPM pada Google Collab.
![NW Google](https://github.com/user-attachments/assets/ffe90fb6-a81d-4fba-ad94-79f6a06f0296)
_Membuat worksheet baru_
![Input GW](https://github.com/user-attachments/assets/773af25f-9be1-4108-a5a6-6490067b21ce)
_Menginputkan Nama dan NPM kemudian ditampilkan_

3.  Membuat akun di https://[www.kaggle.com/.](http://www.kaggle.com/)
![Kaggle](https://github.com/user-attachments/assets/7e9c8981-a69e-42d0-977c-dd6328d16997)

4.  Membuat akun di https://github.com/.
![Github](https://github.com/user-attachments/assets/8b3449d6-e308-4b47-89a8-088717aab481)

5.  Melakukan praktek dari https://youtu.be/mSO2hJln0OY?feature=shared. Praktek tersebut yaitu:
    5.1 Load sample dataset
![5 1 Load Dataset](https://github.com/user-attachments/assets/5b819a12-89d0-40c0-abb0-32f77933fc75)
    5.2 Metadata | Deskripsi dari sample dataset
![Metadata](https://github.com/user-attachments/assets/dc648f7e-3cdb-44dc-9b7e-718a72017a33)
    5.3 Explanatory & Response Variables | Features & Target
![Explanatory](https://github.com/user-attachments/assets/50b9f306-aad4-4149-8379-03d3f6ce15a7)
   5.4  Feature & Target Names
![Feature](https://github.com/user-attachments/assets/671abee1-5fe6-40f0-93de-d9b98590ec10)
  5.5   Visualisasi Data
![Visualisasi](https://github.com/user-attachments/assets/0486a5d0-b787-408b-bcbb-92f39bdcde48)
  5.6	 Training Set & Testing Set
![Training](https://github.com/user-attachments/assets/b2c76e72-ba1b-41e4-b568-e81e04ca71b7)
  5.7  Load sample dataset sebagai Pandas Data Frame
![Load Sample dataset](https://github.com/user-attachments/assets/59212bee-f9cc-4572-90a6-316e34d1aa06)

6.  Melakukan praktek dari https://youtu.be/tiREcHrtDLo?feature=shared. Praktek tersebut yaitu:
     6.1 Persiapan dataset | Loading & splitting dataset
![6 1](https://github.com/user-attachments/assets/da1708df-5e28-4014-b0c2-c94058e3d08d)
     6.2  Training model Machine Learning
![6 2](https://github.com/user-attachments/assets/3fba24f2-a8d7-4d9f-bcba-d9c80192a337)
    6.3 Evaluasi model Machine Learning
![6 3](https://github.com/user-attachments/assets/d6b5fdd7-bd73-4e14-a3a5-3fb5288fcfce)
    6.4 Pemanfaatan trained model machine learning
![6 4](https://github.com/user-attachments/assets/4e30626c-966e-4042-8d5c-2ec9fad7366e)
    6.5 Deploy model Machine Learning | Dumping dan Loading model Machine Learning
![6 5 dumping](https://github.com/user-attachments/assets/35ef9c70-fdf9-4516-a06b-7e7d504b533b)
_Proses dumping_ 
![6 5 load](https://github.com/user-attachments/assets/802e303e-eb92-41a0-bcc3-9c21d3814efe)
_Proses loading_
![6 5 jika dumping](https://github.com/user-attachments/assets/73b198d9-60e5-4c2a-8a77-32b7649c93df)
_Jika dumping berhasil dilakukan maka akan secara otomatis membuat file baru bernama iris_classifier_kkn.joblib._

7.  Melakukan praktek dari https://youtu.be/smNnhEd26Ek?feature=shared. Praktek tersebut yaitu:
     7.1 Persiapan sample dataset
![7 1](https://github.com/user-attachments/assets/83370cc4-8150-43fd-b7d5-349a166f81f2)
    7.2 Teknik data preprocessing 1: binarisation
![7 2](https://github.com/user-attachments/assets/6d5262a0-da07-4733-9d22-802255735dd1)
   7.3 Teknik data preprocessing 2: scaling
![7 3](https://github.com/user-attachments/assets/09302796-5ec3-429b-9467-63496a81e4df)
   7.4 Teknik data preprocessing 3: normalization
![7 4 l1](https://github.com/user-attachments/assets/ed75acc5-8994-44de-a00e-7bd2c48a3768)
_L1 Normalisation: Lest Absolute Deviations_
![7 4 l2](https://github.com/user-attachments/assets/c09783cc-f29d-4694-8b85-0b37c35464bf)
_L2 Normalisation: Least Squares_

**Comment by nessa0712** [comment]
### TUGAS 3
**Decision Tree, Random Forrest.**

**1. Decision Tree Classification
    1.1 Pengenalan komponen Decision Tree: root, node, leaf**
![1 1](https://github.com/user-attachments/assets/e89ce2be-a734-4099-acda-7ad345da2616)
         Decision Tree adalah model klasifikasi yang menerapkan struktur pohon. Dalam decision tree biasanya memilki 3 
         komponen penting, yaitu :
     •	Root Node: Node tertinggi dalam struktur pohon atau struktur data pohon. Node akar merupakan elemen global yang mewakili seluruh pesan dan tidak memiliki induk.
    •	Internal Node: Simpul atau node dalam pohon yang memiliki anak dan bukan daun.
    •	Leaf Node: simpul yang berada pada tingkat paling bawah dari pohon dalam struktur data, sehingga tidak memiliki anak. Setiap leaf dari decision tree akan mempresentasikan prediksi class yaang akan dihasilkan dari struktur tree tersebut.
    **1.2  Pengenalan Gini Impurity**
Gini Impurity adalah fungsi kerugian yang menggambarkan kemungkinan salah klasifikasi untuk satu sampel, menurut distribusi sekumpulan data berlabel tertentu.
    **1.3  Pengenalan Information Gain**
Information Gain merupakan proses mengukur pengurangan entropi atau kejutan dengan membagi kumpulan data menurut nilai tertentu dari variabel acak.
    **1.4  Membangun Decision Tree**
          • Membuat keputusan
         • Masukkan beberapa pertimbangan untuk mengambil keputusan
         • Melakukan pengujian
         • Membuat kesimpulan
     **1.5  Persiapan dataset: Iris Dataset**
    ![1 5](https://github.com/user-attachments/assets/c3281083-da39-4cfa-a0e9-ca9961a2aec7)
    **1.6  Training model Decision Tree Classifier**
![1 6](https://github.com/user-attachments/assets/9519fdef-2559-4e6a-a779-fca55508e6e4)
    **1.7 Visualisasi model Decision Tree**
![1 7](https://github.com/user-attachments/assets/736d3023-0438-4b8c-afc6-6f1d6d881c0b)
   **1.8 Evaluasi model Decision Tree**
![1 8](https://github.com/user-attachments/assets/81f95c5a-5d9c-4e33-b893-02314aabbdfe)

**2. Random Forest Classification
    2.1 Proses training model Machine Learning secara umum**
![2 1](https://github.com/user-attachments/assets/02d68282-8e72-4764-9592-b8340d907f40)
Proses training selalu diawali dengan training set yang berisikan kumpulan model atau biasa di presentasikan sebagai X_train dan target model y_train. Training set ini digunakan untuk melakukan training model machine learning dan nantinya model yang sudah di training ini dinamakan trained model. Selanjutnya trained model ini akan digunakan untuk melakukan prediksi terhadap sekumpulan nilai futures yang baru. Kemudian, hasil prediksi akan di presentasikan sebagai y_pred.
  **2.2 Pengenalan Ensemble Learning**
![2 2](https://github.com/user-attachments/assets/5b7866c5-76ca-44f8-ae70-edd027f275b7)
Ensemble Learning adalah teknik yang dikenal di machine learning dimana menggangabungkan beberapa model untuk melakukan prediksi. Pada Ensemble Learning diawali dengan training set namun, training set akan digunakan untuk mentraining beberapa model (KNN, SVM, dan Decision Tree) dan akan menghasilkan tiga trained model. Selanjutnya trained model akan digunakan untuk melakukan prediksi terhadapat sekumupulan nilai futrures yang baru. Karena menghasilkan tiga prediksi maka diperlukan pengumpulan data yang dinamakan mean or mode.
**2.3 Pengenalan Bootstrap Aggregating | Bagging**
![2 3](https://github.com/user-attachments/assets/1132bdae-583a-4c92-89b5-06a1f4bfad6a)
Homogenuous Ensemble Learning adalah teknik yang memadukan sejumlah machine learning yang sejenis. Sama dengan teknik lainnya teknik ini juga diawali dengan training set. Bootstrap Aggregating atau Bagging adalah teknik yang menerapkan proses random sampling with replacement terhadap training set yang kita miliki dan akan menghasilkan beberapa training set yang baru sejumlah model yang kita training. Training set baru yang dihasilkan dari proses bagging ini dikenal dengan kasus bag. Dan bag ini berisikan kumpulan data yang berbada antara satu dengan yang lainnya. Kemudian bag ini akan digunakan untuk mentraining model machie learning yang sejenis.
**2.4 Pengenalan Random Forest | Hutan Acak**
![2 4](https://github.com/user-attachments/assets/c1ed1471-52d8-4139-87f5-91e4bc2e5b2f)
Random forest merupakan resendpretasi dari Homogenuous Ensemble Learning yang menerapkan decision tree. Random forest tercipta dari sekumpulan dari decision tree. Karena random forest ini pada awalnya perkumupulan decision tree. Dalam implemetasi nya selain bagging random forest juga mengaktifkan features randomness dimana untuk setiap bag yang dihasilkan akan mengadopsi sejumlah features yang diadobsi dari training set sumbernya. Dan nantinya decision akan di training dari data set yang berbeda maka akan dihasilkan trained model yang berbeda-beda. Selanjutnya trained model ini akan digunakan untuk membuat prediksi terhadap nilai features yang baru dan pengumpulan data akan dilakukan mean or mode. Strategi ini membuat Random forest lebih baik dibandingkan model decision tree yang lain karena setiap tree yang ada di training menggunakan subset dari keseluruhan training set yang ada.
**2.5 Persiapan dataset | Iris Flower Dataset**
![2 5](https://github.com/user-attachments/assets/84f68a38-fd55-496a-a01b-ff600aeedb9a)
**2.6 Implementasi Random Forest Classifier dengan Scikit Learn**
![2 6](https://github.com/user-attachments/assets/859216f8-cde6-4616-b95d-03f69bccba01)
**2.7 Evaluasi model dengan Classification Report**
![2 7](https://github.com/user-attachments/assets/929b732d-d416-4b44-b9b7-f8b70f628f19)

**Comment by nessa0712** [comment]
### TUGAS 4 
**Naive Bayes, Confusion Matrix.**

1. Navive Bayes Classification
1.1.	Pengenalan Bayes Theorem | Teori Bayes | Conditional Probability
Bayes' theorem merupakan suatu formula yang menawarkan untuk menghitung nilai probability dari suatu event dengan memanfaatkan pengetahuan sebelumnya dari kondisi terkait atau sering kali dikenal dengan istilah conditional probability.
![WhatsApp Image 2024-11-22 at 11 12 11](https://github.com/user-attachments/assets/b853a57f-25f1-4153-a24a-54dae108a4e8)

1.2.	Pengenalan Naive Bayes Classification
Naive Bayes Classifier merupakan sebuah metoda klasifikasi dengan menggunakan metode probabilitas dan statistik yang berakar pada teorema Bayes yang ditemukan oleh ilmuwan Inggris Thomas Bayes. Ciri utama dari Naive Bayes Classifier ini adalah asumsi yg sangat kuat akan independensi dari masing-masing kondisi atau kejadian.

1.3.	Pengenalan Prior Probability
Prior adalah nilai probability dari kemunculan target label nilai tertentu tanpa memperhatikan niali feature nya.

1.4.	Pengenalan Likelihood
Likelihood dalam naive bayes adalah probability nilai kemunculan dari features tertentu bila diketahui nilai kemuculan nilai target label nya.

1.5.	Pengenalan Evidence | Normalizer
Evidence dalam naive bayes adalah total akumulasi dari hasil perkalian antara nilai Likelihood dengan nilai prior nya.

1.6.	Pengenalan Posterior Probability
Probability dalam naive bayes adalah probabilitas yang diperbarui dari suatu peristiwa setelah mempertimbangkan informasi baru. Probabilitas posterior dihitung dengan memperbarui probabilitas sebelumnya menggunakan teorema Bayes.

1.7.	Persiapan Dataset | Wiscoinsin Breadt Cancer Dataset | UCI Machine Learning
![1 7](https://github.com/user-attachments/assets/4233b7b7-0a42-471e-a7b9-7b5c984cf6c7)
_Persiapan Dataset_
![1 7 info](https://github.com/user-attachments/assets/b8485198-294d-49de-ad46-57fddbe69773)
_Informasi mengenai fuction jupyter notebook yang ada di jupyter notebook_
![1 7 persiapan](https://github.com/user-attachments/assets/9d53529c-5fed-4d5f-9a0b-5459ef3be89f)
_Persiapan Dataset_
![1 7 training](https://github.com/user-attachments/assets/442c7d60-479f-4a33-8622-8d2463ebf1b0)
_Training dan Testing Dataset_

1.8 Implementasi Naïve Bayes Classification dengan Scikit-Learn
![1 8](https://github.com/user-attachments/assets/a8e8bb1a-c430-4afa-92ae-f4c22ce7f44c)
